File: clients/ios/JarvisClient/JarvisClient/Services/GRPCVoiceService.swift
// GRPCVoiceService.swift
// Jarvis iOS Client
//
// Manages the full lifecycle of the VoiceService.Converse bidirectional gRPC
// stream. This is the single point of contact between the iOS app and the
// Jarvis voice backend — AudioCaptureEngine feeds audio in, VoiceViewModel
// consumes events out.
//
// Dependencies (via Swift Package Manager):
//   - grpc-swift       → https://github.com/grpc/grpc-swift  (>= 2.0.0)
//   - swift-protobuf   → https://github.com/apple/swift-protobuf (>= 1.26.0)
//
// Generated stubs expected at:
//   JarvisClient/Proto/voice.pb.swift        (swift-protobuf)
//   JarvisClient/Proto/voice.grpc.swift      (grpc-swift)

import Foundation
import GRPC
import NIOCore
import NIOPosix
import SwiftProtobuf
import os.log

// MARK: - Public Event Types

/// All downstream events the app can receive from the Jarvis voice backend.
/// Delivered on the MainActor via the `onEvent` callback.
public enum VoiceEvent {
    case statusChanged(JarvisStatusEvent.State, message: String)
    case transcriptReceived(text: String, isFinal: Bool, confidence: Float)
    case replyReceived(replyText: String, intent: String, requiresConfirmation: Bool)
    case audioReplyReceived(data: Data, text: String, isFinalChunk: Bool)
    case hudActionReceived(JarvisHUDAction)
    case streamError(VoiceServiceError)
    case streamClosed
}

/// Errors surfaced by GRPCVoiceService.
public enum VoiceServiceError: Error, LocalizedError {
    case notConnected
    case sessionCapacityExceeded
    case streamSetupFailed(underlying: Error)
    case sendFailed(underlying: Error)
    case reconnectLimitReached(attempts: Int)
    case invalidConfiguration(String)

    public var errorDescription: String? {
        switch self {
        case .notConnected:
            return "Voice stream is not connected."
        case .sessionCapacityExceeded:
            return "Jarvis backend has reached its session limit."
        case .streamSetupFailed(let err):
            return "Stream setup failed: \(err.localizedDescription)"
        case .sendFailed(let err):
            return "Failed to send audio: \(err.localizedDescription)"
        case .reconnectLimitReached(let n):
            return "Reconnect failed after \(n) attempts."
        case .invalidConfiguration(let msg):
            return "Invalid configuration: \(msg)"
        }
    }
}

// MARK: - Configuration

/// All tuneable parameters for GRPCVoiceService.
public struct VoiceServiceConfiguration {
    /// gRPC server host, e.g. "jarvis.yourdomain.com" or "localhost".
    public var host: String

    /// gRPC server port. Typically 443 (TLS) or 50059 (dev plaintext).
    public var port: Int

    /// Use TLS. Always true in production; false for local docker-compose dev.
    public var useTLS: Bool

    /// BCP-47 language code forwarded to the backend STT engine.
    public var languageCode: String

    /// Optional voice ID for TTS reply synthesis.
    public var voiceID: String

    /// NLP context tags forwarded to nlp-service ParseIntent per utterance.
    public var contextTags: [String]

    /// Maximum number of automatic reconnect attempts on stream failure.
    public var maxReconnectAttempts: Int

    /// Base delay for exponential backoff reconnect (doubles each attempt).
    public var reconnectBaseDelay: TimeInterval

    /// NIO thread count for the gRPC channel's event loop group.
    public var nioThreads: Int

    public static var development: VoiceServiceConfiguration {
        VoiceServiceConfiguration(
            host: "localhost",
            port: 50059,
            useTLS: false,
            languageCode: "en-US",
            voiceID: "",
            contextTags: [],
            maxReconnectAttempts: 5,
            reconnectBaseDelay: 0.5,
            nioThreads: 1
        )
    }

    public static var production: VoiceServiceConfiguration {
        VoiceServiceConfiguration(
            host: "voice.jarvis.yourdomain.com",
            port: 443,
            useTLS: true,
            languageCode: "en-US",
            voiceID: "",
            contextTags: [],
            maxReconnectAttempts: 8,
            reconnectBaseDelay: 1.0,
            nioThreads: 2
        )
    }
}

// MARK: - GRPCVoiceService

/// Manages the bidirectional gRPC stream for VoiceService.Converse.
///
/// Usage:
/// ```swift
/// let service = GRPCVoiceService(configuration: .development) { event in
///     // handle VoiceEvent on the main actor
/// }
/// await service.connect(userID: "tony", sessionID: UUID().uuidString)
/// await service.sendAudioChunk(data: pcmData, capturedAt: .now, isWakeWordFrame: false)
/// await service.sendEndOfSpeech()
/// await service.disconnect()
/// ```
@MainActor
public final class GRPCVoiceService: ObservableObject {

    // MARK: - Published State

    @Published public private(set) var isConnected: Bool = false
    @Published public private(set) var currentState: JarvisStatusEvent.State = .idle
    @Published public private(set) var reconnectAttempt: Int = 0

    // MARK: - Private

    private let config: VoiceServiceConfiguration
    private let onEvent: (VoiceEvent) -> Void
    private let log = Logger(subsystem: "com.jarvis.client", category: "GRPCVoiceService")

    /// NIO event loop group shared across reconnects.
    private var group: MultiThreadedEventLoopGroup?

    /// gRPC channel — recreated on reconnect.
    private var channel: GRPCChannel?

    /// The active bidirectional call handle.
    private var call: BidirectionalStreamingCall<JarvisConverseRequest, JarvisConverseResponse>?

    /// Active session metadata set on connect.
    private var sessionID: String = ""
    private var userID: String = ""

    /// Sequence counter for outgoing AudioChunks.
    private var sequenceNum: Int64 = 0

    /// Guards against concurrent connect/disconnect calls.
    private var isConnecting: Bool = false

    // MARK: - Init

    public init(
        configuration: VoiceServiceConfiguration,
        onEvent: @escaping (VoiceEvent) -> Void
    ) {
        self.config = configuration
        self.onEvent = onEvent
    }

    deinit {
        // Best-effort shutdown — channel and group are reference types
        // so no await needed here; teardown happens asynchronously.
        try? channel?.close().wait()
        try? group?.syncShutdownGracefully()
    }
}

// MARK: - Public Interface

extension GRPCVoiceService {

    /// Opens the gRPC channel, dials the server, and sends the StreamConfig
    /// as the very first message. Safe to call again after a disconnect.
    public func connect(userID: String, sessionID: String) async {
        guard !isConnecting && !isConnected else { return }
        isConnecting = true
        defer { isConnecting = false }

        self.userID    = userID
        self.sessionID = sessionID
        self.sequenceNum = 0

        do {
            try await openChannelAndStream()
            isConnected = true
            reconnectAttempt = 0
            log.info("Connected — session: \(sessionID)")
        } catch {
            log.error("Connection failed: \(error.localizedDescription)")
            onEvent(.streamError(.streamSetupFailed(underlying: error)))
        }
    }

    /// Sends a raw PCM audio chunk upstream.
    /// Call this from AudioCaptureEngine's frame callback.
    public func sendAudioChunk(
        data: Data,
        capturedAt: Date = .now,
        isWakeWordFrame: Bool = false
    ) async {
        sequenceNum += 1
        let chunk = JarvisAudioChunk.with {
            $0.data             = data
            $0.sequenceNum      = sequenceNum
            $0.capturedAtMs     = Int64(capturedAt.timeIntervalSince1970 * 1000)
            $0.isWakeWordFrame  = isWakeWordFrame
        }
        await send(.with { $0.audio = chunk })
    }

    /// Signals end-of-speech. Triggers the STT → NLP → TTS pipeline on the server.
    public func sendEndOfSpeech() async {
        await sendControlEvent(.endOfSpeech)
    }

    /// Cancels the current in-flight pipeline without closing the stream.
    public func sendCancel() async {
        await sendControlEvent(.cancel)
    }

    /// Sends a keep-alive ping to prevent stream timeout during silence.
    public func sendKeepAlive() async {
        await sendControlEvent(.keepAlive)
    }

    /// Signals the start of a new conversation turn (e.g. PTT re-pressed).
    public func sendNewTurn() async {
        await sendControlEvent(.newTurn)
    }

    /// Gracefully closes the stream and channel.
    public func disconnect() async {
        log.info("Disconnecting — session: \(sessionID)")
        await teardown()
        onEvent(.streamClosed)
    }
}

// MARK: - Internal: Channel + Stream Setup

private extension GRPCVoiceService {

    func openChannelAndStream() async throws {
        // ── 1. NIO Event Loop Group ─────────────────────────────────────
        let elg = MultiThreadedEventLoopGroup(numberOfThreads: config.nioThreads)
        self.group = elg

        // ── 2. gRPC Channel ─────────────────────────────────────────────
        var channelBuilder = ClientConnection
            .usingPlatformAppropriateTLS(for: elg)

        if !config.useTLS {
            channelBuilder = ClientConnection.insecure(group: elg)
        }

        let channel = channelBuilder
            .withBackgroundActivityLogger(makeGRPCLogger())
            .connect(host: config.host, port: config.port)

        self.channel = channel

        // ── 3. Stub ─────────────────────────────────────────────────────
        let stub = JarvisVoiceServiceNIOClient(channel: channel)

        // ── 4. Open bidirectional stream ─────────────────────────────────
        let call = stub.converse(
            callOptions: makeCallOptions(),
            handler: { [weak self] response in
                // Responses arrive on a NIO thread — hop to MainActor.
                Task { @MainActor [weak self] in
                    self?.handleResponse(response)
                }
            }
        )
        self.call = call

        // ── 5. Send StreamConfig as first message ────────────────────────
        let configMsg = JarvisConverseRequest.with {
            $0.config = makeStreamConfig()
        }
        try await call.sendMessage(configMsg).get()

        // ── 6. Observe stream completion for reconnect logic ─────────────
        Task { [weak self] in
            do {
                // Awaits until the server closes the stream or an error occurs.
                _ = try await call.status.get()
                await MainActor.run { [weak self] in
                    guard let self else { return }
                    self.isConnected = false
                    self.log.info("Stream closed cleanly")
                    self.onEvent(.streamClosed)
                }
            } catch {
                await MainActor.run { [weak self] in
                    guard let self else { return }
                    self.isConnected = false
                    self.log.error("Stream error: \(error.localizedDescription)")
                    self.onEvent(.streamError(.streamSetupFailed(underlying: error)))
                    Task { await self.attemptReconnect() }
                }
            }
        }
    }

    // MARK: StreamConfig factory

    func makeStreamConfig() -> JarvisStreamConfig {
        JarvisStreamConfig.with {
            $0.meta = JarvisRequestMeta.with {
                $0.requestID = UUID().uuidString
                $0.userID    = userID
                $0.sessionID = sessionID
                $0.source    = "voice"
            }
            $0.clientInfo = JarvisClientInfo.with {
                $0.platform    = "ios"
                $0.appVersion  = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown"
                $0.deviceModel = UIDevice.current.model
                $0.osVersion   = UIDevice.current.systemVersion
            }
            $0.audioConfig = JarvisAudioConfig.with {
                $0.encoding        = .pcm16Bit
                $0.sampleRateHz    = 16000
                $0.channelCount    = 1
                $0.frameDurationMs = 20
            }
            $0.languageCode = config.languageCode
            $0.voiceID      = config.voiceID
            $0.contextTags  = config.contextTags
        }
    }
}

// MARK: - Internal: Response Handling

private extension GRPCVoiceService {

    /// Dispatches a VoiceResponse to the appropriate VoiceEvent.
    /// Always called on the MainActor.
    func handleResponse(_ response: JarvisConverseResponse) {
        switch response.payload {

        case .status(let s):
            currentState = s.state
            onEvent(.statusChanged(s.state, message: s.message))

        case .transcript(let t):
            onEvent(.transcriptReceived(
                text:       t.text,
                isFinal:    t.isFinal,
                confidence: t.confidence
            ))

        case .reply(let r):
            onEvent(.replyReceived(
                replyText:            r.replyText,
                intent:               r.intent,
                requiresConfirmation: r.requiresConfirmation
            ))

        case .audioReply(let a):
            onEvent(.audioReplyReceived(
                data:         a.data,
                text:         a.text,
                isFinalChunk: a.isFinalChunk
            ))

        case .action(let a):
            onEvent(.hudActionReceived(a))

        case .none:
            log.warning("Received VoiceResponse with no payload")
        }
    }
}

// MARK: - Internal: Send Helpers

private extension GRPCVoiceService {

    func send(_ request: JarvisConverseRequest) async {
        guard let call else {
            onEvent(.streamError(.notConnected))
            return
        }
        do {
            try await call.sendMessage(request).get()
        } catch {
            log.error("sendMessage failed: \(error.localizedDescription)")
            onEvent(.streamError(.sendFailed(underlying: error)))
        }
    }

    func sendControlEvent(_ type: JarvisControlEvent.TypeEnum) async {
        let event = JarvisConverseRequest.with {
            $0.event = JarvisControlEvent.with { $0.type = type }
        }
        await send(event)
    }
}

// MARK: - Internal: Reconnect

private extension GRPCVoiceService {

    /// Exponential backoff reconnect. Preserves the original userID/sessionID
    /// so the server-side session store can stitch the stream back together.
    func attemptReconnect() async {
        guard reconnectAttempt < config.maxReconnectAttempts else {
            log.error("Reconnect limit reached (\(config.maxReconnectAttempts) attempts)")
            onEvent(.streamError(.reconnectLimitReached(attempts: config.maxReconnectAttempts)))
            return
        }

        reconnectAttempt += 1
        let delay = config.reconnectBaseDelay * pow(2.0, Double(reconnectAttempt - 1))
        log.info("Reconnect attempt \(reconnectAttempt) in \(String(format: "%.1f", delay))s")

        try? await Task.sleep(nanoseconds: UInt64(delay * 1_000_000_000))

        await teardown()
        await connect(userID: userID, sessionID: sessionID)
    }
}

// MARK: - Internal: Teardown

private extension GRPCVoiceService {

    func teardown() async {
        // Half-close the send side (tells the server the client is done sending).
        try? await call?.sendEnd().get()
        call = nil

        // Close the channel and shut down NIO.
        try? await channel?.close().get()
        channel = nil

        try? group?.syncShutdownGracefully()
        group = nil

        isConnected = false
    }
}

// MARK: - Internal: gRPC Utilities

private extension GRPCVoiceService {

    func makeCallOptions() -> CallOptions {
        var options = CallOptions()
        // Auth token injected here — replace with your token provider.
        // options.customMetadata.add(name: "authorization", value: "Bearer \(token)")
        options.timeLimit = .none   // streaming — no deadline
        return options
    }

    func makeGRPCLogger() -> Logger {
        Logger(subsystem: "com.jarvis.client", category: "GRPC")
    }
}

File: clients/android/app/src/main/kotlin/com/jarvis/client/grpc/GRPCVoiceService.kt
// GRPCVoiceService.kt
// Jarvis Android Client — grpc/
//
// Manages the full lifecycle of the VoiceService.Converse bidirectional gRPC
// stream. Single point of contact between the Android app and the Jarvis
// voice backend — AudioCaptureEngine feeds audio in, VoiceViewModel
// consumes events out via StateFlow / SharedFlow.
//
// Threading:
//   connect/disconnect/send* must be called from a coroutine (any dispatcher).
//   gRPC callbacks arrive on gRPC executor threads and are forwarded via flows.
//   VoiceViewModel collects on Dispatchers.Main.
//
// Reconnect strategy:
//   Exponential backoff, capped at maxReconnectAttempts. The original
//   userID + sessionID are preserved so the server session store can stitch
//   the stream back together (especially valuable with the Redis backend).

package com.jarvis.client.grpc

import android.os.Build
import android.util.Log
import com.jarvis.client.audio.AudioFrame
import io.grpc.ManagedChannel
import io.grpc.ManagedChannelBuilder
import io.grpc.Status
import io.grpc.stub.StreamObserver
import jarvis.voice.Voice.*
import jarvis.common.Common.*
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*
import java.util.UUID
import java.util.concurrent.TimeUnit
import kotlin.math.pow

private const val TAG = "GRPCVoiceService"

// ── Public Event Types ────────────────────────────────────────────────────────

/** All downstream events the app can receive from the Jarvis voice backend. */
sealed class VoiceEvent {
    data class StatusChanged(val state: StatusEvent.State, val message: String) : VoiceEvent()
    data class TranscriptReceived(val text: String, val isFinal: Boolean, val confidence: Float) : VoiceEvent()
    data class ReplyReceived(val replyText: String, val intent: String, val requiresConfirmation: Boolean) : VoiceEvent()
    data class AudioReplyReceived(val data: ByteArray, val text: String, val isFinalChunk: Boolean) : VoiceEvent() {
        override fun equals(other: Any?) = other is AudioReplyReceived && data.contentEquals(other.data)
        override fun hashCode() = data.contentHashCode()
    }
    data class HudActionReceived(val action: HUDAction) : VoiceEvent()
    data class StreamError(val error: VoiceServiceError) : VoiceEvent()
    data object StreamClosed : VoiceEvent()
}

sealed class VoiceServiceError : Exception() {
    data object NotConnected : VoiceServiceError()
    data object SessionCapacityExceeded : VoiceServiceError()
    class StreamSetupFailed(cause: Throwable) : VoiceServiceError()
    class SendFailed(cause: Throwable) : VoiceServiceError()
    data class ReconnectLimitReached(val attempts: Int) : VoiceServiceError()
    data class InvalidConfiguration(val detail: String) : VoiceServiceError()
}

// ── Configuration ─────────────────────────────────────────────────────────────

data class VoiceServiceConfiguration(
    val host:                 String  = "localhost",
    val port:                 Int     = 50059,
    val useTls:               Boolean = false,
    val languageCode:         String  = "en-US",
    val voiceId:              String  = "",
    val contextTags:          List<String> = emptyList(),
    val maxReconnectAttempts: Int     = 5,
    val reconnectBaseDelaySec: Double = 0.5,
    val appVersion:           String  = "1.0.0",
) {
    companion object {
        val Development = VoiceServiceConfiguration()
        val Production  = VoiceServiceConfiguration(
            host                  = "voice.jarvis.yourdomain.com",
            port                  = 443,
            useTls                = true,
            maxReconnectAttempts  = 8,
            reconnectBaseDelaySec = 1.0,
        )
    }
}

// ── GRPCVoiceService ──────────────────────────────────────────────────────────

class GRPCVoiceService(
    private val config: VoiceServiceConfiguration,
    private val scope:  CoroutineScope,
) {

    // ── State ─────────────────────────────────────────────────────────────────

    private val _isConnected = MutableStateFlow(false)
    val isConnected: StateFlow<Boolean> = _isConnected.asStateFlow()

    private val _currentState = MutableStateFlow(StatusEvent.State.STATE_IDLE)
    val currentState: StateFlow<StatusEvent.State> = _currentState.asStateFlow()

    private val _reconnectAttempt = MutableStateFlow(0)
    val reconnectAttempt: StateFlow<Int> = _reconnectAttempt.asStateFlow()

    /** Hot flow of all downstream VoiceEvent messages. */
    private val _events = MutableSharedFlow<VoiceEvent>(extraBufferCapacity = 32)
    val events: SharedFlow<VoiceEvent> = _events.asSharedFlow()

    // ── Private ───────────────────────────────────────────────────────────────

    private var channel:        ManagedChannel? = null
    private var requestStream:  StreamObserver<ConverseRequest>? = null
    private var sessionId:      String = ""
    private var userId:         String = ""
    private var sequenceNum:    Long   = 0L

    @Volatile private var isConnecting = false

    // ── Public Interface ──────────────────────────────────────────────────────

    /** Opens the gRPC channel and sends the StreamConfig first message. */
    suspend fun connect(userId: String, sessionId: String) {
        if (isConnecting || _isConnected.value) return
        isConnecting = true
        try {
            this.userId    = userId
            this.sessionId = sessionId
            this.sequenceNum = 0

            openChannelAndStream()
            _isConnected.value    = true
            _reconnectAttempt.value = 0
            Log.i(TAG, "Connected — session: $sessionId")
        } catch (e: Exception) {
            Log.e(TAG, "Connection failed: ${e.message}")
            emit(VoiceEvent.StreamError(VoiceServiceError.StreamSetupFailed(e)))
        } finally {
            isConnecting = false
        }
    }

    /** Sends a raw PCM audio chunk upstream. */
    suspend fun sendAudioChunk(frame: AudioFrame, isWakeWordFrame: Boolean = false) {
        sequenceNum++
        val chunk = audioChunk {
            data            = com.google.protobuf.ByteString.copyFrom(frame.pcmData)
            sequenceNum     = this@GRPCVoiceService.sequenceNum
            capturedAtMs    = frame.capturedAtMs
            isWakeWordFrame = isWakeWordFrame
        }
        send(converseRequest { audio = chunk })
    }

    /** Signals end-of-speech — triggers STT → NLP → TTS on the server. */
    suspend fun sendEndOfSpeech()  = sendControlEvent(ControlEvent.Type.TYPE_END_OF_SPEECH)

    /** Cancels the current in-flight pipeline without closing the stream. */
    suspend fun sendCancel()       = sendControlEvent(ControlEvent.Type.TYPE_CANCEL)

    /** Sends a keep-alive ping to prevent idle stream timeout. */
    suspend fun sendKeepAlive()    = sendControlEvent(ControlEvent.Type.TYPE_KEEP_ALIVE)

    /** Signals the start of a new conversation turn. */
    suspend fun sendNewTurn()      = sendControlEvent(ControlEvent.Type.TYPE_NEW_TURN)

    /** Gracefully closes the stream and channel. */
    suspend fun disconnect() {
        Log.i(TAG, "Disconnecting — session: $sessionId")
        teardown()
        emit(VoiceEvent.StreamClosed)
    }

    // ── Private: Channel + Stream Setup ──────────────────────────────────────

    private fun openChannelAndStream() {
        val ch = buildChannel()
        channel = ch

        // Response observer — dispatches incoming messages to [_events].
        val responseObserver = object : StreamObserver<ConverseResponse> {
            override fun onNext(response: ConverseResponse) {
                handleResponse(response)
            }
            override fun onError(t: Throwable) {
                val grpcStatus = Status.fromThrowable(t)
                Log.e(TAG, "Stream error: ${grpcStatus.code} — ${grpcStatus.description}")
                _isConnected.value = false
                emit(VoiceEvent.StreamError(VoiceServiceError.StreamSetupFailed(t)))
                scope.launch { attemptReconnect() }
            }
            override fun onCompleted() {
                Log.i(TAG, "Stream completed by server")
                _isConnected.value = false
                emit(VoiceEvent.StreamClosed)
            }
        }

        // Bidirectional streaming requires the Java async stub — gRPC-Kotlin's
        // coroutine stub exposes server/client streaming separately, not bidi.
        val asyncStub = jarvis.voice.VoiceServiceGrpc.newStub(ch)
        requestStream = asyncStub.converse(responseObserver)

        // Send StreamConfig as the very first message.
        requestStream?.onNext(converseRequest { config = makeStreamConfig() })
            ?: throw IllegalStateException("requestStream is null after creation")
    }

    private fun buildChannel(): ManagedChannel {
        val builder = if (config.useTls) {
            ManagedChannelBuilder.forAddress(config.host, config.port)
        } else {
            ManagedChannelBuilder.forAddress(config.host, config.port)
                .usePlaintext()
        }
        return builder
            .keepAliveTime(30, TimeUnit.SECONDS)
            .keepAliveTimeout(10, TimeUnit.SECONDS)
            .build()
    }

    private fun makeStreamConfig(): StreamConfig = streamConfig {
        meta = requestMeta {
            requestId = UUID.randomUUID().toString()
            userId    = this@GRPCVoiceService.userId
            sessionId = this@GRPCVoiceService.sessionId
            source    = "voice"
        }
        clientInfo = clientInfo {
            platform    = "android"
            appVersion  = config.appVersion
            deviceModel = "${Build.MANUFACTURER} ${Build.MODEL}"
            osVersion   = Build.VERSION.RELEASE
        }
        audioConfig = audioConfig {
            encoding       = AudioEncoding.AUDIO_ENCODING_PCM_16BIT
            sampleRateHz   = 16_000
            channelCount   = 1
            frameDurationMs = 20
        }
        languageCode = config.languageCode
        voiceId      = config.voiceId
        contextTags.addAll(config.contextTags)
    }

    // ── Private: Response Handling ─────────────────────────────────────────────

    private fun handleResponse(response: ConverseResponse) {
        when (response.payloadCase) {
            ConverseResponse.PayloadCase.STATUS -> {
                val s = response.status
                _currentState.value = s.state
                emit(VoiceEvent.StatusChanged(s.state, s.message))
            }
            ConverseResponse.PayloadCase.TRANSCRIPT -> {
                val t = response.transcript
                emit(VoiceEvent.TranscriptReceived(t.text, t.isFinal, t.confidence))
            }
            ConverseResponse.PayloadCase.REPLY -> {
                val r = response.reply
                emit(VoiceEvent.ReplyReceived(r.replyText, r.intent, r.requiresConfirmation))
            }
            ConverseResponse.PayloadCase.AUDIO_REPLY -> {
                val a = response.audioReply
                emit(VoiceEvent.AudioReplyReceived(a.data.toByteArray(), a.text, a.isFinalChunk))
            }
            ConverseResponse.PayloadCase.ACTION -> {
                emit(VoiceEvent.HudActionReceived(response.action))
            }
            else -> Log.w(TAG, "Received ConverseResponse with no payload")
        }
    }

    // ── Private: Send Helpers ─────────────────────────────────────────────────

    private suspend fun send(request: ConverseRequest) {
        val stream = requestStream
        if (stream == null) {
            emit(VoiceEvent.StreamError(VoiceServiceError.NotConnected))
            return
        }
        try {
            withContext(Dispatchers.IO) { stream.onNext(request) }
        } catch (e: Exception) {
            Log.e(TAG, "sendMessage failed: ${e.message}")
            emit(VoiceEvent.StreamError(VoiceServiceError.SendFailed(e)))
        }
    }

    private suspend fun sendControlEvent(type: ControlEvent.Type) {
        send(converseRequest {
            event = controlEvent { this.type = type }
        })
    }

    // ── Private: Reconnect ────────────────────────────────────────────────────

    private suspend fun attemptReconnect() {
        val attempt = _reconnectAttempt.value + 1
        if (attempt > config.maxReconnectAttempts) {
            Log.e(TAG, "Reconnect limit reached (${config.maxReconnectAttempts} attempts)")
            emit(VoiceEvent.StreamError(VoiceServiceError.ReconnectLimitReached(config.maxReconnectAttempts)))
            return
        }
        _reconnectAttempt.value = attempt

        val delayMs = (config.reconnectBaseDelaySec * 1_000 * 2.0.pow(attempt - 1)).toLong()
        Log.i(TAG, "Reconnect attempt $attempt in ${delayMs}ms")
        delay(delayMs)

        teardown()
        connect(userId, sessionId)
    }

    // ── Private: Teardown ─────────────────────────────────────────────────────

    private suspend fun teardown() {
        withContext(Dispatchers.IO) {
            try { requestStream?.onCompleted() } catch (_: Exception) {}
            requestStream = null
            try {
                channel?.shutdown()
                channel?.awaitTermination(3, TimeUnit.SECONDS)
            } catch (_: Exception) {}
            channel = null
        }
        _isConnected.value = false
    }

    // ── Private: Flow helper ─────────────────────────────────────────────────

    private fun emit(event: VoiceEvent) {
        _events.tryEmit(event)
    }
}

// ── Protobuf DSL builders ─────────────────────────────────────────────────────
// These top-level functions let us write `converseRequest { ... }` instead of
// `ConverseRequest.newBuilder().apply { ... }.build()`.

private fun converseRequest(block: ConverseRequest.Builder.() -> Unit): ConverseRequest =
    ConverseRequest.newBuilder().apply(block).build()

private fun streamConfig(block: StreamConfig.Builder.() -> Unit): StreamConfig =
    StreamConfig.newBuilder().apply(block).build()

private fun requestMeta(block: RequestMeta.Builder.() -> Unit): RequestMeta =
    RequestMeta.newBuilder().apply(block).build()

private fun clientInfo(block: ClientInfo.Builder.() -> Unit): ClientInfo =
    ClientInfo.newBuilder().apply(block).build()

private fun audioConfig(block: AudioConfig.Builder.() -> Unit): AudioConfig =
    AudioConfig.newBuilder().apply(block).build()

private fun audioChunk(block: AudioChunk.Builder.() -> Unit): AudioChunk =
    AudioChunk.newBuilder().apply(block).build()

private fun controlEvent(block: ControlEvent.Builder.() -> Unit): ControlEvent =
    ControlEvent.newBuilder().apply(block).build()
